### PR TITLE
Add `--benchmark` support.

### PR DESCRIPTION
**Pull request by pq**
New benchmark mode that prints lint stats based on the best of 10 runs.

Follow-up from: https://github.com/dart-lang/linter/issues/669

@devoncarew @bwilkerson

### CODE REVIEW

**Comment by pq** [comment]
FYI @alexeieleusis : even with the best of, it looks like `invariant_booleans` is running a bit slow... ?

**Comment by alexeieleusis** [comment]
What time is it taking?

**Comment by pq** [comment]
144 ms (from [here](https://travis-ci.org/dart-lang/linter/jobs/234151797))

**Comment by alexeieleusis** [comment]
I'll look into it once this is merged, at least we have a baseline now.

File: bin/linter.dart [hunk 1]
```diff
@@ -54,6 +55,8 @@ void runLinter(List<String> args, LinterOptions initialLintOptions) {
         abbr: "h", negatable: false, help: "Show usage information.")
     ..addFlag("stats",
         abbr: "s", negatable: false, help: "Show lint statistics.")
+    ..addFlag("benchmark",
```
Comment (devoncarew) [inline bin/linter.dart]: may not need a shorthand for this flag, given how rarely it'll be used?

File: bin/linter.dart [hunk 2]
```diff
@@ -141,21 +144,27 @@ void runLinter(List<String> args, LinterOptions initialLintOptions) {
   }
 
   var stats = options['stats'];
-  if (stats == true) {
+  var benchmark = options['benchmark'];
+  if (stats == true || benchmark == true) {
```
Comment (devoncarew) [inline bin/linter.dart]: perhaps `(stats || benchmark)`?

File: lib/src/formatter.dart [hunk 3]
```diff
@@ -291,6 +266,57 @@ class SimpleFormatter implements ReportFormatter {
   }
 }
 
+writeBenchmarks(IOSink out, List<File> filesToLint, LinterOptions lintOptions) {
```
Comment (devoncarew) [inline lib/src/formatter.dart]: so few return types :(

File: tool/travis.sh [hunk 4]
```diff
@@ -20,8 +20,8 @@ if [ "$LINTER_BOT" = "benchmark" ]; then
   # The actual lints can have errors - we don't want to fail the benchmark bot.
   set +e
 
-  # Run linter with all lints enabled.
-  dart bin/linter.dart -s -q -c example/all.yaml .
+  # Benchmark linter with all lints enabled.
+  dart bin/linter.dart -b -q -c example/all.yaml .
```
Comment (devoncarew) [inline tool/travis.sh]: `--benchmark` if you remove the shorthand

File: bin/linter.dart (hunk 1 above)
Comment (pq) [inline bin/linter.dart]: Sounds good.

File: bin/linter.dart (hunk 2 above)
Comment (pq) [inline bin/linter.dart]: Yes!  For some reason I thought I convinced myself that undefined would return `null` but yeah, much better. 👍

File: lib/src/formatter.dart (hunk 3 above)
Comment (pq) [inline lib/src/formatter.dart]: Yep.  Consider this bit of code my last stand against gratuitous returns (in non API bits). 😸

File: tool/travis.sh (hunk 4 above)
Comment (pq) [inline tool/travis.sh]: Done.

**Comment by pq** [comment]
@devoncarew : take another gander?

**Comment by devoncarew** [comment]
lgtm!

**Comment by pq** [comment]
Thanks!

**Pull request merged**